### PR TITLE
Bad operationId to error status code

### PR DESCRIPTION
**Pull request by jfinkhaeuser**
Adds the ability to turn bad or missing operationIds into run time errors returning a defined status code, e.g. 501 ("Not Implemented").

This change is largely useful for testing purposes.

When (black-box) testing implementations, it is not yet know whether a given implementation matches the given Swagger/OpenAPI specs. Especially during development, partial implementations might exist and test well, but some endpoints might still be missing from the implementation.  This change lets connexion serve such partial implementations, and lets testers write expectations about missing functionality (i.e. results in a 501 error).

Note that this PR includes #274, #275, #276 and #277, and in fact depends on those changes. I would recommend merging the others first (or not, as it were) before merging this one.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -3,7 +3,7 @@ max-line-length=120
 exclude=connexion/__init__.py
 
 [tox]
-envlist=pypy,py27,py34,py35,py35-flask0.10.1,isort-check,flake8
+envlist=py27,py34
```
Comment (rafaelcaricio) [inline tox.ini]: Please change this back. Not sure if you committed that change by mistake.

**Comment by rafaelcaricio** [comment]
Rebase this PR please.

File: tox.ini (hunk 1 above)
Comment (jfinkhaeuser) [inline tox.ini]: Oh yes, very much so! Sorry about that!

**Comment by jfinkhaeuser** [comment]
Yep. Probably tomorrow.

File: tests/test_api.py [hunk 2]
```diff
@@ -59,11 +63,28 @@ def test_invalid_operation_does_not_stop_application_in_debug_mode():
...
+    api = Api(TEST_FOLDER / "fixtures/module_not_implemented/swagger.yaml", "/api/v1.0",
+              {'title': 'OK'}, debug=True)
+    assert api.specification['info']['title'] == 'OK'
+
     api = Api(TEST_FOLDER / "fixtures/user_module_loading_error/swagger.yaml", "/api/v1.0",
               {'title': 'OK'}, debug=True)
     assert api.specification['info']['title'] == 'OK'
 
-    api = Api(TEST_FOLDER / "fixtures/module_not_implemented/swagger.yaml", "/api/v1.0",
+    api = Api(TEST_FOLDER / "fakeapi/missing_op_id.yaml", "/api/v1.0",
```
Comment (rafaelcaricio) [inline tests/test_api.py]: Please use this swagger file from the `/fixtures` dir.

**Comment by rafaelcaricio** [comment]
@jfinkhaeuser Would be nicer to have this as an option, it could be called `stub_missing_operations` (or something similar) so it would be aligned with what I tried to do in https://github.com/zalando/connexion/pull/284. In my PR I used 400 status code which not I see would be nice to be 501. I am willing to change the #284 to use this PR.

**Comment by jfinkhaeuser** [comment]
Makes sense as an option to me, yes.

Whether you change #284 or I change this one, I don't mind particularly :)

**Comment by rafaelcaricio** [comment]
It is preferably to this change to be in this PR since it is where this behavior is going to be merged.

**Comment by jfinkhaeuser** [comment]
Alright, I'll look into it.

File: tests/test_api.py [hunk 3]
```diff
@@ -59,11 +63,28 @@ def test_invalid_operation_does_not_stop_application_in_debug_mode():
...
+
+def test_other_errors_stop_application_to_setup():
+    # The previous tests were just about operationId not being resolvable.
+    # Other errors should still result exceptions!
+    with pytest.raises(InvalidSpecification):
+        Api(TEST_FOLDER / "fakeapi/bad_specs.yaml", "/api/v1.0",
+            {'title': 'OK'})
+
+    # Debug mode should ignore the error
+    api = Api(TEST_FOLDER / "fakeapi/bad_specs.yaml", "/api/v1.0",
```
Comment (rafaelcaricio) [inline tests/test_api.py]: Could you please move those specs to a proper directory within `/fixtures`?

**Comment by jfinkhaeuser** [comment]
So now I've re-familiarized myself with everything, I've got to ask: nice to have as an option to what?

Right now, it's an option to `add_api`, i.e. if `resolver_error` is an int, that's the status code that'll be used. If none, the default behaviour is used.

It's relatively hard adding this functionality to a `StubResolver`  as in #284. There's a bunch of information missing in `Resolver` to make that easy, and IMHO it's not a resolver's purpose to make up things that didn't resolve.

So I've put it into Api instead, which makes a lot more sense to me. `add_api` gets the option for convenience.

File: tests/test_api.py (hunk 3 above)
Comment (jfinkhaeuser) [inline tests/test_api.py]: Ah, yeah, the bad specs I missed. One moment.

**Comment by rafaelcaricio** [comment]
Yes, I looked through your code here after posting my comment and I see that it is already an option. I will do a final review before merging it.

File: connexion/app.py [hunk 4]
```diff
@@ -96,7 +96,7 @@ def common_error_handler(exception):
 
     def add_api(self, swagger_file, base_path=None, arguments=None, auth_all_paths=None, swagger_json=None,
                 swagger_ui=None, swagger_path=None, swagger_url=None, validate_responses=False,
-                strict_validation=False, resolver=Resolver()):
+                strict_validation=False, resolver=Resolver(), resolver_error=None):
```
Comment (rafaelcaricio) [inline connexion/app.py]: It is not very clear that `resolver_error` is expected to be a HTTP status code. Would be nice to have a more friendly name here.

**Comment by rafaelcaricio** [comment]
👍

**Pull request merged**